# Landscape of plant mitochondrial genome via dis-assembly approach.



## Short Description



This repository is for my research on plant mitochondrial genomes, which I call "mitogenome" in this README.md. Shortly and roughly speaking, mitogenomes of plants are different from those of animals or fungus. For example, 

- The sizes are much larger.
- Genomes contain some repetitive regions, and
-  (repeat-)mediated recombinations create divergent structures called "multipartite structure"s.

I'm trying to reconstruct these multipartite structures in the plant mitogenomes. In contrast to usual assembly workflows, I'd call my approach as "dis-"assembly, as I leverage existing reference genomes and "find" other structures by clustering.



## Properties

- Author: Bansho Masutani @ University of Tokyo
- Contact: ban-m@g.ecc.u-tokyo.ac.jp


## Data Availability

The following list is the list of accessions/URLs to the data used in this research:

- Seven accessions of Arabidopsis thaliana via [Sequence Read Archive]( https://www.ebi.ac.uk/ena/data/view/PRJEB31147)
- One Ler accession from [PacBio's official repository](https://downloads.pacbcloud.com/public/SequelData/ArabidopsisDemoData/)
- One Col-0 accession sequenced newly in the National Institute of Genetics in Japan (BioSample: SAMD00224135-SAMD00224139).

All of these data were generated by Sequel Systems. For more details, e.g., the used chemistory, please visit the links.

The Gen Bank IDs of the reference sequences used in this research are BK010421.1, JF729100, and JF729102. BK010421.1 is the reference sequence I'm using generally.


WGS reads were filtered to enrich reads from mitochondrial genome. Specifically, we run the following command:`bash ./script/filter_reads ${WGS} ${BK010421.1} > ${FILTERED}`.


The filtered reads, the outpot of PO-HMM, and supplementary plots are available at the following locations. "Circos," "linear," and "no-marge" represent a circos plot (like in the paper), a dot plot between assembled contig and the reference, and a circos plots without any "aggressive merging," respectively.

|Strain Name | Reads| Result | Circos | Linear | no_merge |
|:----------:|:----:|:------:|:------:|:------:|:--------:|
|pacbio|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/pacbio.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/pacbio.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/pacbio/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/pacbio/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/pacbio/no_merge.html)|
|ler|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/ler.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/ler.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/ler/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/ler/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/ler/no_merge.html)|
|col0_1106_exp2|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/col0_1106_exp2.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/col0_1106_exp2.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/col0_1106_exp2/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/col0_1106_exp2/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/col0_1106_exp2/no_merge.html)|
|cvi|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/cvi.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/cvi.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/cvi/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/cvi/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/cvi/no_merge.html)|
|an1|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/an1.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/an1.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/an1/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/an1/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/an1/no_merge.html)|
|c24|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/c24.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/c24.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/c24/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/c24/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/c24/no_merge.html)|
|kyo|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/kyo.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/kyo.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/kyo/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/kyo/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/kyo/no_merge.html)|
|sha|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/sha.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/sha.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/sha/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/sha/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/sha/no_merge.html)|
|eri|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/filtered_reads/eri.fasta download>Reads</a>|<a href = https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/eri.tar.gz download>Result</a>|[Circos](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/eri/circos.html)|[Liner](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/eri/linear.html)|[NoMerge](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/viewer/eri/no_merge.html)|


## Reproducibility

### Synthetic dataset

First, install the dependencies. Minimum requirements are 

- [BadRead](https://github.com/rrwick/Badread): `git clone https://github.com/rrwick/Badread.git && pip3 install ./Badread` would install `BadRead` to your local envirnment.
- [Flye](https://github.com/fenderglass/Flye)
- [Minimap2](https://github.com/lh3/minimap2): `git clone https://github.com/lh3/minimap2 && cd minimap2 && make` and `ln -s ${PWD}/minimap2 ${HOME}/local/bin/minimap2` or anywhere included in the $PATH variable.
- [Samtools](https://github.com/samtools/samtools/releases/tag/1.10)
- [HTSlib](https://github.com/samtools/htslib/releases/tag/1.10.2)
- [Last](http://last.cbrc.jp/)
- [Rust](https://www.rust-lang.org/)


To obtain baseline result, please install

- [WhatsHap](https://github.com/whatshap/whatshap): `pip3 install --user whatshap`
- [isONclust](https://github.com/ksahlin/isONclust): `pip install isONclust`.
- [CARNAC-LR](https://github.com/kamimrcht/CARNAC-LR): Exec `git clone https://github.com/kamimrcht/CARNAC.git && cd CARNAC && make`. Then, you can sim-link `${PWD}/CARNAC-LR ${HOME}/local/bin/CARNAC-LR` or anywhere included in the $PATH variable. Also, copy `CARNAC/scripts/paf_to_CARNAC.py` into `./script/` of this repository.


Then, 
```bash
git clone --recursive https://github.com/ban-m/reconstruct_mito_genome.git
cd reconstruct_mito_genome
cargo check && cargo build && cargo build --release # Build binaries.
bash ./script/create_mock_genome.sh # create datasets under ${PWD}/data/synthetic_data/ by using `BadRead`
bash ./script/synthetic_dataset.sh
bash ./script/benchmark.sh # create results under ${PWD}/result/benchmark/
bash ./script/posterior_probability.job
bash ./script/baseline_mock_genomes.job # Invoke `isONclust`, `CARNAC-LR`, `WhatsHap`, and `Flye`
```
would create all the result used in the paper. Note that each script would take long time to be done (from a few hours to a few days).


### Real dataset

First, download the dataset and the reference genome.
```bash
bash ${PWD}/donwloads_files.sh
```

Please type `ll -ht ${PWD}/data/filtered_reads` to confirm that all the dataset are properly donwloaded; `md5sum` values are as follows:

- 35ce43718fc0832f55f691424f337d69  NC_037304_1.fa
- e040ac3f8b047b7e84d38cb70faf8ecc  an1.fasta
- 7b4b5f70ba29891a09a83786db5d8ad9  c24.fasta
- 00157130264b868bf35dd10828749a3d  col0_1106_exp2.fasta
- 893394f4ef02fe4737fb26b7896ea98f  cvi.fasta
- 26646b904a44bd564c84ccf5bcc6c7e9  eri.fasta
- 072427917074882e4ad9dbc4b03ad87d  kyo.fasta
- 67b2f656d2eaa91126f36b6b111a665f  ler.fasta
- 4a92ad3d6fe66732e2fcba4d76488b53  pacbio.fasta
- 69a0007d689ba6a797c0e975a0eb8092  sha.fasta

Then, run the pipeline on each dataset. Here, ${threads} is the number of threads to be used. We used ${threads}=24.

```bash
bash ./script/real_dataset.sh ${threads}
```
It takes several days to complete on a 24 threads 2.0GHz computer.

## Instalation

This program depends on following libraries and binaries:

- [HTSlib](https://github.com/samtools/htslib/releases/tag/1.10.2)
- [Last](http://last.cbrc.jp/)
- [Rust](https://www.rust-lang.org/) (Nightly needed)

After install these requirements, do

```bath
git clone --recursive https://github.com/ban-m/reconstruct_mito_genome.git
cd reconstruct_mito_genome
cargo check && cargo build && cargo build --release
```

Then, `target/release/mmmm --help` would print help messages to `stderr`.

Basic usage would be

```
mmmm decompose --output ${OUTPUT} \
    --reads ${READ} --contigs ${REFERENCE} \
    --cluster_num ${MIN_CLUSTER} --threads ${CORES}
```
${READ} and ${REFERENCE} should be fasta files. ${MIN_CLUSTER} is the size of the cluster at each window, not the number of resulting cluster.

*Disclaimer1*: It requires a reference quality contigs for clutering. For those dataset without very good reference, I'm currently developing 
[this repository](https://github.com/ban-m/hla_haplotyper).

*Disclaimer2*: As the clustering algorithm (PO-HMM) is very costly, this program is for very small genome(<1Mbp). 


## Reference

- Jiao, W., Schneeberger, K. Chromosome-level assemblies of multiple *Arabidopsis* genomes reveal hotspots of rearrangements with altered evolutionary dynamics.
  *Nat Commun* **11,** 989 (2020). https://doi.org/10.1038/s41467-020-14779-y

