# Landscape of plant mitochondrial genome via dis-assembly approach.



## Short Description



This repository is for my research on plant mitochondrial genomes, which I call "mitogenome" in this README.md. Shortly and roughly speaking, mitogenomes of plants are different from those of animals or fungus. For example, 

- The sizes are much larger.
- Genomes contain some repetitive regions, and
-  (repeat-)mediated recombinations create divergent structures called "multipartite structure"s.

I'm trying to reconstruct these multipartite structures in the plant mitogenomes. In contrast to usual assembly workflows, I'd call my approach as "dis-"assembly, as I leverage existing reference genomes and "find" other structures by clustering.



## Properties

- Author: Bansho Masutani at University of Tokyo
- Last Update: 2020/05/29
- Contact: ban-m@g.ecc.u-tokyo.ac.jp


## Data Availability

The following list is the list of accessions/URLs to the data used in this research:

- Seven accessions of Arabidopsis thaliana via [Sequence Read Archive]( https://www.ebi.ac.uk/ena/data/view/PRJEB31147)
- One Ler accession from [PacBio's official repository](https://downloads.pacbcloud.com/public/SequelData/ArabidopsisDemoData/)
- One Col-0 accession sequenced newly in the National Institute of Genetics in Japan (BioSample: SAMD00224135-SAMD00224139).

All of these data were generated by Sequel Systems. For more details, e.g., the used chemistory, please visit the links.

The Gen Bank IDs of the reference sequences used in this research are BK010421.1, JF729100, and JF729102. BK010421.1 is the reference sequence I'm using generally.

Also, the supplementary plots are available at the following locations. "Circos," "linear," and "no-marge" represent a circos plot (like in the paper), a dot plot between assembled contig and the reference, and a circos plots without any "aggressive merging," respectively.

|Strain Name | Circos | Linear | no_merge |
|:----------:|:------:|:------:|:--------:|
|pacbio | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/pacbio/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/pacbio/no_merge.html)|
|ler | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/ler/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/ler/no_merge.html)|
|col0| [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/col0_1106_exp2/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/col0_1106_exp2/no_merge.html)|
|sha | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/sha/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/sha/no_merge.html)|
|cvi-0 | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/cvi/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/cvi/no_merge.html)|
|an1 | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/an1/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/an1/no_merge.html)|
|c24 | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/c24/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/c24/no_merge.html)|
|kyoto | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/kyo/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/kyo/no_merge.html)|
|eri-1 | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/eri/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/eri/no_merge.html)|
|pacbio_ler | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/pacbio_ler/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/pacbio_ler/no_merge.html)|
|ler_ler | [View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/ler_ler/circos.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly//linear.html)|[View](https://mlab.cb.k.u-tokyo.ac.jp/~ban-m/mitochondria_assembly/ler_ler/no_merge.html)|

Note: We used BK010421.1 as the reference genome except pacbio_ler and ler_ler where we used JF729100 as the reference genome.

## Reproducibility

### Synthetic dataset

First, install the required packeges:

- [BadRead](https://github.com/rrwick/Badread): `git clone https://github.com/rrwick/Badread.git && pip3 install ./Badread` would install `BadRead` to your local envirnment.
- [isONclust](https://github.com/ksahlin/isONclust): `pip install isONclust`.
- [CARNAC-LR](https://github.com/kamimrcht/CARNAC-LR): Exec `git clone https://github.com/kamimrcht/CARNAC.git && cd CARNAC && make`. Then, you can sim-link `${PWD}/CARNAC-LR ${HOME}/local/bin/CARNAC-LR` or anywhere included in the $PATH variable. Also, copy `CARNAC/scripts/paf_to_CARNAC.py` into `./script/` of this repository.
- [Flye](https://github.com/fenderglass/Flye)
- [Minimap2](https://github.com/lh3/minimap2): `git clone https://github.com/lh3/minimap2 && cd minimap2 && make` and `ln -s ${PWD}/minimap2 ${HOME}/local/bin/minimap2` or anywhere included in the $PATH variable.
- [WhatsHap](https://github.com/whatshap/whatshap): `pip3 install --user whatshap`
- [Samtools](https://github.com/samtools/samtools/releases/tag/1.10)
- [HTSlib](https://github.com/samtools/htslib/releases/tag/1.10.2)
- [Last](http://last.cbrc.jp/)
- [Rust](https://www.rust-lang.org/)

Then, 
```bash
git clone --recursive https://github.com/ban-m/reconstruct_mito_genome.git
cd reconstruct_mito_genome
cargo check && cargo build && cargo build --release # Build binaries.
bash ./script/create_mock_genome.sh # create datasets under ${PWD}/data/synthetic_data/ by using `BadRead`
bash ./script/benchmark.sh # create results under ${PWD}/result/benchmark/
bash ./script/posterior_probability.job
bash ./script/synthetic_dataset.sh
bash ./script/baseline_mock_genomes.job # Invoke `isONclust`, `CARNAC-LR`, `WhatsHap`, and `Flye`
```
would create all the result used in the paper. Note that each script would take long time to be done (from a few hours to a few days).


### Real dataset

First, download the dataset.
```bash
```

Then, run the pipeline on each dataset.
```bash
bash ./script/real_dataset.sh ${threads}
```
It takes several days to complete on a 24 threads 2.0GHz computer.

## Requirements and Instalation

This program depends on following libraries and binaries:

- [HTSlib](https://github.com/samtools/htslib/releases/tag/1.10.2)
- [Last](http://last.cbrc.jp/)
- [Rust](https://www.rust-lang.org/) (Nightly needed)

After install these requirements, exec

```bath
git clone --recursive https://github.com/ban-m/reconstruct_mito_genome.git
cd reconstruct_mito_genome
cargo check && cargo build && cargo build --release
```

Then, `target/release/mmmm --help` would print help messages to `stderr`.

Basic usage would be
```mmmm decompose --output ${OUTPUT} \
    --reads ${READ} --contigs ${REFERENCE} \
    --cluster_num ${MIN_CLUSTER} --threads ${CORES}
```
${READ} and ${REFERENCE} should be fasta files. ${MIN_CLUSTER} is the size of the cluster at each window, not the number of resulting cluster.

*Disclaimer1*: It requires a reference quality contigs for clutering. For those dataset without very good reference, I'm currently developing 
[this repository](https://github.com/ban-m/hla_haplotyper).

*Disclaimer2*: As the clustering algorithm (PO-HMM) is very costly, this program is for very small genome(<1Mbp). 


## Reference

- Jiao, W., Schneeberger, K. Chromosome-level assemblies of multiple *Arabidopsis* genomes reveal hotspots of rearrangements with altered evolutionary dynamics.
  *Nat Commun* **11,** 989 (2020). https://doi.org/10.1038/s41467-020-14779-y

